modules/ln_core/utf8string.scm

```diff
@@ -177,53 +177,84 @@ SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
-            ((fx= #xE0 (bitwise-and (car l) #xF0))
-                (cons (bitwise-ior (arithmetic-shift (bitwise-and (list-ref l 0) #x0F) 12)
-                                   (arithmetic-shift (bitwise-and (list-ref l 1) #x3F)  6)
-                                   (bitwise-and (list-ref l 2) #x3F))
-                      (decode (list-tail l 3))))
-            ((fx= #xF0 (bitwise-and (car l) #xF8))
-                (cons (bitwise-ior (arithmetic-shift (bitwise-and (list-ref l 0) #x07) 18)
-                                   (arithmetic-shift (bitwise-and (list-ref l 1) #x3F) 12)
-                                   (arithmetic-shift (bitwise-and (list-ref l 2) #x3F)  6)
-                                   (bitwise-and (list-ref l 3) #x3F))
-                      (decode (list-tail l 4))))
-            ((fx= #xF8 (bitwise-and (car l) #xFC))
-                (cons (bitwise-ior (arithmetic-shift (bitwise-and (list-ref l 0) #x03) 24)
-                                   (arithmetic-shift (bitwise-and (list-ref l 1) #x3F) 18)
-                                   (arithmetic-shift (bitwise-and (list-ref l 2) #x3F) 12)
-                                   (arithmetic-shift (bitwise-and (list-ref l 3) #x3F)  6)
-                                   (bitwise-and (list-ref l 4) #x3F))
-                      (decode (list-tail l 5))))
-            ((fx= #xFC (bitwise-and (car l) #xFE))
-                (cons (bitwise-ior (arithmetic-shift (bitwise-and (list-ref l 0) #x01) 30)
-                                   (arithmetic-shift (bitwise-and (list-ref l 1) #x3F) 24)
-                                   (arithmetic-shift (bitwise-and (list-ref l 2) #x3F) 18)
-                                   (arithmetic-shift (bitwise-and (list-ref l 3) #x3F) 12)
-                                   (arithmetic-shift (bitwise-and (list-ref l 4) #x3F)  6)
-                                   (bitwise-and (list-ref l 5) #x3F))
-                      (decode (list-tail l 6)))))))
-     (decode (map char->integer (string->list src))))
+
+(define utf8string->unicode:on-encoding-error
+  (let ((handler #f))
+    (define (utf8string->unicode:on-encoding-error-replace str idx) #xfffd)
+    (define (utf8string->unicode:on-encoding-error-raise-error str idx)
+      (error "UTF-8 char encoding error" str idx))
+    (define (utf8string->unicode:on-encoding-error-log+replace str idx)
+      (log-error "UTF-8 char encoding error" str idx)
+      #xfffd)
+    (define (select key)
+      (set!
+       handler
+       (if (procedure? key)
+           key
+           (case key
+             ((replace #f) utf8string->unicode:on-encoding-error-replace)
+             ((log+replace) utf8string->unicode:on-encoding-error-log+replace)
+             (else utf8string->unicode:on-encoding-error-raise-error)))))
+    (select 'error)
+    (lambda args
+      (if (null? args) handler (select (car args))))))
+
+(define (utf8string->unicode str #!optional (encoding-error #t))
```

Comment:
Can't say I fully understand this piece, but in fairness probably also not the previous version of it. However, as we don't have any unit tests on this file, how would I go about evaluating that it does as it should? @ddunsmuir uses more interesting language apps that I do, but ideally we don't want to break those only to learn something didn't go as planned?

Reply:
> Can't say I fully understand this piece, but in fairness probably also not the previous version of it.

Do you refer t o the definition of `utf8string->unicode:on-encoding-error` or the `utf8-string->unicode`?

The former is just a setter, which takes either no arguments and returns the current handler, of takes the first arg and if it's a procedure trusts the caller and sets it as handler or is selects a pre-defined handler based on the symbol passed in.

The latter is just a slightly modified copy (the encoding error handler is new) of the code I'm using for years with utf-8.  And I do not have many things which still work when UTF-8 is broken.

> However, as we don't have any unit tests on this file, how would I go about evaluating that it does as it should?

So far I feed the old implementation with some broken UTF-8 (accidentally - I typed a "ä" into a glgui-label) which did not handle it.  The new does.  Otherwise it should rather clearly implement the URF-8 decoding.  IMHO this is a case, where it would be better to validate the source correctness than try to come up with bound to be incomplete test cases.

The error handling is another thing.  For now it just get's the job done better than the old one, because it enables users to somehow continue to work with lambdanative.  The predefined handlers seem to be compatible with the a) the old behavior, just better message b) a presumably useful case, which would log when it replaces characters and the replacing version, which at least keeps the app running.

> @ddunsmuir uses more interesting language apps that I do, but ideally we don't want to break those only to learn something didn't go as planned?